### ISSUE TITLE
SPEC: allow types to be specified for declarations

### ISSUE THREAD

**Issue by jsternberg** [influxdata/ifql#344]
I think we should allow types to be declared in the specification. While the language does not have to specify types, types do exist and it is statically typed. I've thought a lot about types and I personally think we should allow optional type specifications in ifql. I don't think we'll do that, but I would like to open up the possibility of alternative languages that do have user-specified function types.

To do this, I feel that it has to be included in the spec. As an example, imagine that we had a function specification like this:

```
type FunctionParam struct {
    Key         *Identifier `json:"key"`
    Type        *Type       `json:"type"`
    Default     Expression  `json:"default"`
    Piped       bool        `json:"piped,omitempty"`
    declaration VariableDeclaration
}
```

If a type is not specified for a function parameter, then it is inferred. But, this would allow a person to implement a function specification that says, "I am expecting this function to take a specific type of argument".

I also think this is related to polymorphic function signatures and function overloading that is implied as needed in the "Function types" section. I'm going to open another issue for addressing that specifically though.

**Comment by nathanielc** [comment]
My goal is for IFQL is to have all types inferred in any context.

My plan is to use a [Hindley-Milner](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system) type system, which by formal specification supports type inference.  The basic idea behind the system is that each function has free type variables. For example

```
add = (a,b) => a + b
x = add(a:1.5, b:2)
```

This function has three free variables in its signature: the type of `a`, the type of `b` and the return type, lets call those variables `A`, `B` and `R` respectively. The type variables are variable because we do not yet know their type. The process of type inference is to solve for the concrete type of each type variable. We can use information available within the program to determine the which of the many types each type variable could be. Since the `+` operator is only defined for `int`, `uint`, `float` and `string` types the type variables `A`, `B` and `R` can only be one of those types. Since we can also see that `add` is called with a float type on the second line, we can infer that `A` must be a float type. Since the `+` operator for floats is defined only for other float and returns a float we can infer that `B` and `R` must also be floats. Once we have solved the type for each "free type variable" we have completed type inference.

That is my understanding of how type inference works. As for how we could implement that, we can leverage the built in functions and operators to infer the types of everything in an IFQL script. My hope is that a user would never need to explicitly declare a type within the IFQL script, because we can always infer the correct type. Specifically this means that built in types, may specify their function signatures more precisely using a mixture of both concrete types and free type variables. Then for every function defined in pure IFQL we will assume that each argument has its own free type variable. With the information from the built in types we can then solve for all free type variables from within IFQL scripts. This will eliminate the need to declare types explicitly as part of the language.

**Comment by jsternberg** [comment]
I need to do more research into this because it's been awhile since I played with Haskell, but I remember Haskell functions (which is the type inference system you reference) act like templates or polymorphism in some regards. Type inference also doesn't make it so there's no need to have types in the underlying system. I also have to note that most systems that do type inference also have a way to specify types in the syntax itself and it's generally recommended as a good practice for libraries to ensure the function is self-documenting and the compiler can give better error messages. Here's an example from a good Haskell beginner's book with the recommendation: http://learnyouahaskell.com/types-and-typeclasses

> Functions also have types. When writing our own functions, we can choose to give them an explicit type declaration. This is generally considered to be good practice except when writing very short functions. From here on, we'll give all the functions that we make type declarations. Remember the list comprehension we made previously that filters a string so that only caps remain? Here's how it looks like with a type declaration.

If I remember correctly, Haskell also has a concept of typeclasses which isn't something that we have and typeclasses would be necessary to implement the kind of behavior you are referencing. I think typeclasses are mathematically a different concept than templates -- templates aren't a mathematical concept -- but are semantically similar. You gave an example of the `+`, but the way `+` is implemented is that each type declares itself to implement the `+` operator using its own types. When the compiler finds a specific usage of `+`, it reifies the actual implementation based on the input types. C++ does the same or a similar thing through templates where it reifies an implementation using a matching template.

The end result is that you still have an underlying type system that needs to be defined at some point and that type inference usually relies on some form of templating and I think we should ensure that is part of the spec so that the query engine can put together the plan.

But, basic polymorphism can also be helpful for some other things. We currently have a `from()` command that accepts two different mutually exclusive parameters. Rather than the extra difficulty of ensuring the `from()` function is used correctly, we could take advantage of a polymorphic system that just has two different `from()` implementations. That way, if someone uses the function incorrectly, it's much easier to display the potential usages to them by saying, "We couldn't find a version of this function that took these arguments with these types. Here are the ones we found." Then instead of having two different implementations inside the single from command, you would have two different implementations that were both registered with the `from()` keyword and were selected based on which types made sense.

Some of these ideas aren't fleshed out and I'm not sure which of the ideas are a function of the language and which ones would be implemented as part of the query language though. I'm more just trying to write out some ideas to get a conversation and to figure out if this is an avenue we should explore. I would like to flesh out the ideas a bit more.
